### PR TITLE
Flex Charge: Add support for TPV store

### PR DESCRIPTION
**Pull request by edgarv09**
Flex Charge: Add support for the TPV store

Test summary:
Local:
5898 tests, 79569 assertions, 0 failures, 17 errors, 0 pendings, 0 omissions, 0 notifications 99.7118% passed
Unit:
12 tests, 58 assertions, 0 failures, 0 errors, 0 pendings, 0 omissions, 0 notifications 100% passed
Remote:
13 tests, 34 assertions, 1 failures, 0 errors, 0 pendings, 1 omissions, 0 notifications 91.6667% passed

### CODE REVIEW

File: lib/active_merchant/billing/gateways/flex_charge.rb [hunk 1]
```diff
@@ -115,23 +135,29 @@ def add_invoice(post, money, credit_card, options)
           avsResultCode: options[:avs_result_code],
           cvvResultCode: options[:cvv_result_code],
           cavvResultCode: options[:cavv_result_code],
-          cardNotPresent: credit_card.verification_value.blank?
+          cardNotPresent: credit_card.is_a?(String) ? false : credit_card.verification_value.blank?
         }.compact
       end
 
       def add_payment_method(post, credit_card, address, options)
```
Comment (gasb150) [inline lib/active_merchant/billing/gateways/flex_charge.rb:142]: The credit_card name could be something more generic, like payment_method considering this is not always a credit_card

File: test/unit/gateways/flex_charge_test.rb [hunk 2]
```diff
@@ -153,6 +153,15 @@ def test_address_names_from_credit_card
     assert_equal 'Doe', names.last
   end
 
+  def test_successful_store
+    response = stub_comms do
+      @gateway.store(@credit_card, @options)
```
Comment (gasb150) [inline test/unit/gateways/flex_charge_test.rb:158]: You can check the request construction with the check_request.

something like 
```ruby
 response = stub_comms do
      @gateway.store(@credit_card, @options)
      end.check_request do |endpoint, data, _headers|
      if endpoint == endpoint_for_access_token_response
       #some assertions
       elsif endpoint == enpoint_for_store_response
       # some assertions
       end
    end.respond_with(successful_access_token_response, successful_store_response)
```

Or even ignore the responses if this approach with two responses is creating some issues.

File: test/remote/gateways/remote_flex_charge_test.rb [hunk 3]
```diff
@@ -147,6 +147,14 @@ def test_failed_fetch_access_token
     assert_match(/400/, error.message)
   end
 
+  def test_successful_purchase_with_token
+    store = @gateway.store(@credit_card_cit, {})
+    assert_success store
+
```
Comment (gasb150) [inline test/remote/gateways/remote_flex_charge_test.rb:153]: You could include assertions representing the store transaction (like a typical response field). (If it's something possible)

File: lib/active_merchant/billing/gateways/flex_charge.rb [hunk 4]
```diff
@@ -115,23 +135,29 @@ def add_invoice(post, money, credit_card, options)
           avsResultCode: options[:avs_result_code],
           cvvResultCode: options[:cvv_result_code],
           cavvResultCode: options[:cavv_result_code],
-          cardNotPresent: credit_card.verification_value.blank?
+          cardNotPresent: credit_card.is_a?(String) ? false : credit_card.verification_value.blank?
```
Comment (gasb150) [inline lib/active_merchant/billing/gateways/flex_charge.rb:138]: Just confirming:  for TPV the cardNotPresent value will always be false.

But for credit cards, will it be something true if no verification value is provided?

**Review by gasb150 — COMMENTED** [review]
@edgarv09 It's pretty good. I left a few comments to your consideration.

File: test/unit/gateways/flex_charge_test.rb (hunk 2 above)
Comment (edgarv09) [reply test/unit/gateways/flex_charge_test.rb:158]: The validations you propose are already carried out in other tests, from my point of view in this case they do not add extra value.

File: test/remote/gateways/remote_flex_charge_test.rb (hunk 3 above)
Comment (edgarv09) [reply test/remote/gateways/remote_flex_charge_test.rb:153]: The only thing in the response that allows us to identify is a store. It's the token and that would be used to set the authorization parameter of the response. So, if this field is not present, blank, or wrong the purchase will fail.

File: lib/active_merchant/billing/gateways/flex_charge.rb (hunk 4 above)
Comment (edgarv09) [reply lib/active_merchant/billing/gateways/flex_charge.rb:138]: Flex charge Documentation is not clear on this topic but the messages during the testing led me to this conclusion.

File: lib/active_merchant/billing/gateways/flex_charge.rb (hunk 1 above)
Comment (edgarv09) [reply lib/active_merchant/billing/gateways/flex_charge.rb:142]: I saw that, but @Heavyblade is still working on flex charge so changing this argument led to merging problems that we are not willing to deal with at this moment.

File: test/unit/gateways/flex_charge_test.rb (hunk 2 above)
Comment (gasb150) [reply test/unit/gateways/flex_charge_test.rb:163]: So the main purpose of this test is the authorization value generated?

File: lib/active_merchant/billing/gateways/flex_charge.rb [hunk 5]
```diff
@@ -228,8 +254,8 @@ def message_from(response)
         response[:title] || response[:responseMessage] || response[:status]
       end
 
-      def authorization_from(response)
-        response[:orderSessionKey]
+      def authorization_from(action, response)
+        action == :store ? response[:transaction][:payment_method][:token] : response[:orderSessionKey]
```
Comment (Heavyblade) [inline lib/active_merchant/billing/gateways/flex_charge.rb:258]: 💬 Comment:

Just in case of unexpected scenarios `dig` is a safer way to get the token

**Review by Heavyblade — COMMENTED** [review]
@edgarv09 looks pretty much ok, left a couple comments for your consideration

File: lib/active_merchant/billing/gateways/flex_charge.rb [hunk 6]
```diff
@@ -115,23 +135,29 @@ def add_invoice(post, money, credit_card, options)
...
-          expirationMonth: credit_card.month,
-          expirationYear: credit_card.year,
-          cardBinNumber: credit_card.number[0..5],
-          cardLast4Digits: credit_card.number[-4..-1],
-          cardNumber: credit_card.number,
-          Token: false
-        }.compact
+        payment_method = case credit_card
+                         when String
+                           { token: true, cardNumber: credit_card }
```
[the remark this answers is not in the document]
Comment (edgarv09) [reply lib/active_merchant/billing/gateways/flex_charge.rb:145]: It seems to work with lower/upper case.

File: CHANGELOG [hunk 7]
```diff
@@ -2,6 +2,7 @@
 = ActiveMerchant CHANGELOG
 
 == HEAD
+* FlexCharge: Add support for TPV store [edgarv09] #5120
```
Comment (Heavyblade) [inline CHANGELOG:5]: 💬 Comment, 

This should be a the bottom of the block

**Review by Heavyblade — APPROVED** [review]

**Pull request merged**